### ISSUE TITLE
Testing shared library with multiple vars scripts cross interaction

### ISSUE THREAD

**Issue by mahmoud-ashi** [jenkinsci/JenkinsPipelineUnit#308]
Hi all. I have a shared library which I am trying to write some tests for but having very hard time figuring out how to do so.

Here is the shared library structure and code (simplified).
```
.
|_ vars
|  |_ fs.groovy
|  |_ lib.groovy
|_ src
|_ test
|  |_ ExampleTest.groovy
|_ build.gradle
```

Here is the content of `lib.groovy`

```groovy
void exec(cmd, label="") {
	if(isUnix()) {
		sh script: cmd, label: label
	} else {
		bat script: cmd, label: label
	}
}

return this
```

Here is the content of `fs.groovy`

```groovy
void mkdir(String dirName) {
	String label = "Creating dir '${dirName}'"
	// Note that isUnix() here is the function coming from the https://plugins.jenkins.io/workflow-basic-steps/ plugin
	String cmd = (isUnix() ? "mkdir -p " : "mkdir ") + dirName
	lib.exec(cmd, label)
}

return this
```

Here is the content of `build.gradle`

```groovy
group 'jenkins-shared-library'
version '1.0-SNAPSHOT'

apply plugin: 'groovy'
apply plugin: 'java'
apply plugin: 'idea'
apply plugin: 'maven'

sourceCompatibility = 1.8
targetCompatibility = 1.8

idea {
  module {
    downloadJavadoc = true
    downloadSources = true
  }
}

sourceSets {
  main {
    groovy {
      srcDirs = ['vars', 'src']
    }
  }

  test {
    groovy {
      srcDirs = ['test']
    }
  }
}

repositories {
  jcenter()
  mavenCentral()
  maven { url 'https://repo.jenkins-ci.org/releases/' }
  maven { url 'https://repo.jenkins-ci.org/public/' }
}

dependencies {
  String JENKINS_VERSION = System.getenv('JENKINS_VERSION') ?: '2.249.1'

  // Frameworks and Jenkins core
  implementation group: 'org.jenkins-ci.main', name: 'jenkins-core', version: JENKINS_VERSION
  testImplementation group: 'org.jenkins-ci.main', name: 'jenkins-core', version: JENKINS_VERSION
  testImplementation group: 'org.jenkins-ci.main', name: 'jenkins-war', version: JENKINS_VERSION
  implementation group: 'org.codehaus.groovy', name: 'groovy-all', version: '2.5.13'
  implementation group: 'junit', name: 'junit', version: '4.12'
  testImplementation group: 'junit', name: 'junit', version: '4.12'
//  implementation group: 'com.lesfurets', name: 'jenkins-pipeline-unit', version: '1.3'
  testImplementation group: 'com.lesfurets', name: 'jenkins-pipeline-unit', version: '1.3'

  // Jenkins plugins
  implementation group: 'com.cloudbees', name: 'groovy-cps', version: '1.31'
  implementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-step-api', version: '2.22', ext: 'jar'
  testImplementation group: 'org.jenkins-ci.plugins', name: 'structs', version: '1.10'
  testImplementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-step-api', version: '2.22', ext: 'jar'
  implementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-api', version: '2.40', ext: 'jar'
  testImplementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-api', version: '2.40', ext: 'jar'
  implementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-durable-task-step', version: '2.36', ext: 'jar'
  testImplementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-durable-task-step', version: '2.36', ext: 'jar'
  implementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-cps', version: '2.83', ext: 'jar'
  testImplementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-cps', version: '2.83', ext: 'jar'
  testImplementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-support', version: '3.5'
  testImplementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-job', version: '2.40'
  implementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-job', version: '2.40', ext: 'jar'
  testImplementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-aggregator', version: '2.6'
  testImplementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-cps-global-lib', version: '2.17'
  implementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-basic-steps', version: '2.20', ext: 'jar'
  testImplementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-basic-steps', version: '2.20', ext: 'jar'
  testImplementation group: 'org.jenkins-ci.main', name: 'jenkins-test-harness', version: '2.67', ext: 'jar'
  testImplementation group: 'org.jenkins-ci.main', name: 'jenkins-test-harness-tools', version: '2.2'
  testImplementation group: 'org.jenkins-ci.plugins', name: 'pipeline-utility-steps', version: '2.6.1', ext: 'jar'
  testImplementation group: 'org.jenkins-ci.lib', name: 'dry-run-lib', version: '0.1'

  // TRICKY: The lib folder contains all other plugins *JAR* files
  // if not found in Maven
  implementation fileTree(dir: 'lib', include: ['*.jar'])
}
```

Let's say I want to test the following cases:
* Mock `isUnix()` to return `true` and make sure that calling `fs.mkdir("dummy")` would call the `sh` command
* Mock `isUnix()` to return `false` and make sure that calling `fs.mkdir("dummy")` would call the `bat` command

So far, what I tried is to add the `ExampleTest.groovy` file in the **test** folder in the same shared library repository which contains the following:
```groovy
import com.lesfurets.jenkins.unit.BasePipelineTest
import org.junit.Before
import org.junit.Test

import static com.lesfurets.jenkins.unit.global.lib.LibraryConfiguration.library
import static com.lesfurets.jenkins.unit.global.lib.ProjectSource.projectSource

class ExampleTest extends BasePipelineTest {
    @Override
    @Before
    void setUp() throws Exception {
        super.setUp()

        def library = library().name("common_lib")
                .defaultVersion("<notNeeded>")
                .allowOverride(true)
                .implicit(false)
                .targetPath("<notNeeded>")
                .retriever(projectSource())
                .build()
        helper.registerSharedLibrary(library)
    }

    @Test
    void exampleTest() {
        helper.registerAllowedMethod("isUnix", [], { return true })
        def script = loadScript("vars/fs.groovy")
        script.mkdir("dummy")
        printCallStack()
    }

    @Test
    void exampleTest2() {
        setUp()
        helper.registerAllowedMethod("isUnix", [], { return true })
        runScript("Jenkins_test")
        printCallStack()
    }
}
```

`Jenkins_test` contains the following basic pipeline
```groovy
// My pipelines usually use the following but saw that it is better for testing to use @Library
// library "common_lib"
@Library("common_lib") _

node("abc") {
    fs.mkdir("dummy")
}
```

Running the command `./gradlew clean test` results in both tests failing with following errors:
* exampleTest --> `No such property: lib for class: fs`
* exampleTest2 --> `No signature of method: fs.isUnix() is applicable for argument types: () values: []`

I have no idea what I am doing wrong. I checked many existing example projects but still cannot find a solution. Your support is highly appreciated.

**Comment by nre-ableton** [comment]
So first of all, I think your `build.gradle` can be _much_ simpler. You don't need to depend on the Jenkins libraries just to make a pipeline library. Have a look at [this `build.gradle` file](https://github.com/Ableton/python-pipeline-utils/blob/master/build.gradle) from a JenkinsPipelineUnit-tested library to see what I mean.

That said, I'm not exactly sure why your pipeline tests are not working correctly, but I *think* the cause might be that you need to also register bindings for these singletons. So in other words, in the `setUp` method of `ExampleTest`, you might need `binding.setVariable('fs', new fs())` (and then import `fs.groovy` in the test class.

Personally I find singletons rather difficult to test for this reason. I think the better approach is for libraries to make classes which contain the bulk of the logic, and test those instead. Then your singletons become passthru classes which don't need to be tested. So for your example, here's what that would look like:

`src/com/yourcompany/FS.groovy`:

```groovy
class FS implements Serializable {
  // Jenkinsfile script context, note that all pipeline steps must use this context
  Object script = null

  void mkdir(String dirName) {
    String label = "Creating dir '${dirName}'"
    // Note that isUnix() here is the function coming from the https://plugins.jenkins.io/workflow-basic-steps/ plugin
    String cmd = (script.isUnix() ? "mkdir -p " : "mkdir ") + dirName
    new Lib(script: script).exec(cmd, label)
}
```

`src/com/yourcompany/Lib.groovy`:

```groovy
class Lib implements Serializable {
  // Jenkinsfile script context
  Object script = null

  void exec(String cmd, String label = "") {
    if (script.isUnix()) {
      script.sh(script: cmd, label: label)
    } else {
      script.bat(script: cmd, label: label)
    }
  }
}
```

...and the singleton `vars/fs.groovy` now becomes:

```groovy
import com.yourcompany.FS

void mkdir(String dirName) {
  new FS(script: this).mkdir(dirName)
}
```

As for the unit tests, they would look like this:

```groovy
import com.lesfurets.jenkins.unit.BasePipelineTest
import org.junit.Before
import org.junit.Test

class ExampleTest extends BasePipelineTest {
    Object script

    @Override
    @Before
    void setUp() throws Exception {
        super.setUp()
        this.script = loadScript('resources/EmptyPipeline.groovy')
    }

    @Test
    void exampleTest() {
        helper.registerAllowedMethod("isUnix", [], { return true })
        FS fs = new FS(script: script)
        fs.mkdir('test')
        printCallStack()
    }
}
```

Note the above `resources/EmptyPipeline.groovy`, which is just:

```groovy
return this
```

The advantage of this approach is that the script context is injected into the pipeline library class, which means you don't need to muck about with executing mock pipelines or mocking `@Library` calls.

**Comment by mahmoud-ashi** [comment]
Thanks @nre-ableton!
I did some changes to my tests and gradle files and now the tests are running correctly. I will share what I did here in case someone faces the same issue in the future.

My `build.gradle` now looks simpler and as following thanks to @nre-ableton comment:
```
group 'jenkins-shared-library'
version '1.0-SNAPSHOT'

apply plugin: 'groovy'
apply plugin: 'idea'

sourceCompatibility = 11
targetCompatibility = 11

idea {
    module {
        downloadJavadoc = true
        downloadSources = true
    }
}

sourceSets {
    main {
        groovy {
            srcDirs = ['vars', 'src']
        }
    }

    test {
        groovy {
            srcDirs = ['test']
        }
    }
}

repositories {
    jcenter()
    mavenCentral()
    maven { url 'https://repo.jenkins-ci.org/releases/' }
}

dependencies {
    // Jenkins core
    String JENKINS_VERSION = System.getenv('JENKINS_VERSION') ?: '2.249.1'
    implementation group: 'org.jenkins-ci.main', name: 'jenkins-core', version: JENKINS_VERSION
    testImplementation group: 'org.jenkins-ci.main', name: 'jenkins-core', version: JENKINS_VERSION

    // Test frameworks
    implementation group: 'org.codehaus.groovy', name: 'groovy-all', version: '3.0.6'
    testImplementation group: 'junit', name: 'junit', version: '4.13.1'
    testImplementation group: 'org.hamcrest', name: 'hamcrest-all', version: '1.3'
    testImplementation group: 'com.lesfurets', name: 'jenkins-pipeline-unit', version: '1.8'
    testImplementation group: 'org.mockito', name: 'mockito-core', version: '3.5.13'

    // Jenkins plugins
    implementation group: 'com.cloudbees', name: 'groovy-cps', version: '1.32'
    implementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-step-api', version: '2.22', ext: 'jar'
    implementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-api', version: '2.40', ext: 'jar'
    implementation group: 'org.jenkins-ci.plugins.workflow', name: 'workflow-cps', version: '2.83', ext: 'jar'

    // TRICKY: The lib folder contains all other plugins *JAR* files if not found in Maven
    implementation fileTree(dir: 'lib', include: ['*.jar'])
}
```

Because I'm testing the scripts in the `vars` dir of my shared library directly, I do not need to load the shared library as I will not be having any pipeline that is loading this shared library for testing. Therefore, I did the following in my `setUp` method:
```groovy
@Before
@Override
void setUp() throws Exception {
    super.setUp()
    loadVarsScriptsToBinding()
}

private void loadVarsScriptsToBinding() {
    // Add bindings to all the "vars" scripts
    File varsDir = new File("vars")
    for (File f : varsDir.listFiles()) {
        if (!f.name.endsWith(".groovy")) {
            continue
        }
        String fileName = f.name.replace(".groovy", "")
        Script = loadScript(f.path)
        binding.setVariable(fileName, s)
    }
}
```

With above `setUp` content, I am now able to write tests like following:
```groovy
@Test
void exampleTest() {
    helper.registerAllowedMethod("isUnix", [], { return true })
    binding.getVariable("fs").mkdir("dummy")
    printCallStack()
}
```

One downside to this solution is that it can get slow which I might actually need @nre-ableton's help here. Because we are loading the scripts before each test, I am adding ~12seconds to each test execution for the loading part. I tried overriding the `initHelper` method and calling the `loadVarsScriptsToBinding` in it but this is also running before each test although I removed the `@Before` on the `setUp` method. I also tried creating a static map which holds the script name and its loaded script and later on just add those again to the `binding` but this does not seem to work. Do you know if there is a better place to call the `loadVarsScriptsToBinding` method to load them only once (either once before all test classes or once per test class)? Thanks in advance.

**Comment by nre-ableton** [comment]
Glad to hear that things are running! I have a few thoughts... first, your `build.groovy` file can be even simpler:

```groovy
group 'jenkins-shared-library'
version '1.0-SNAPSHOT'

apply plugin: 'groovy'
apply plugin: 'idea'

sourceCompatibility = 11
targetCompatibility = 11

idea {
    module {
        downloadJavadoc = true
        downloadSources = true
    }
}

sourceSets {
    main {
        groovy {
            srcDirs = ['vars', 'src']
        }
    }

    test {
        groovy {
            srcDirs = ['test']
        }
    }
}

repositories {
    jcenter()
    mavenCentral()
    maven { url 'https://repo.jenkins-ci.org/releases/' }
}

dependencies {
    // Test frameworks
    implementation group: 'org.codehaus.groovy', name: 'groovy-all', version: '3.0.6'
    testImplementation group: 'junit', name: 'junit', version: '4.13.1'
    testImplementation group: 'org.hamcrest', name: 'hamcrest-all', version: '1.3'
    testImplementation group: 'com.lesfurets', name: 'jenkins-pipeline-unit', version: '1.8'
    testImplementation group: 'org.mockito', name: 'mockito-core', version: '3.5.13'
}
```

You don't need to depend on Jenkins JARs to write a pipeline library, except if you are directly referencing their symbols by `import`ing libraries in your Groovy code (which you usually shouldn't have to do).

Regarding the slowness of the test files, I guess this is because `loadScript` is taking some time. I recommend not testing singletons at all and only testing the actual classes (as shown in my previous post), which means that `loadScript` wouldn't be called at all. I understand that maybe in some cases this is not desirable for whatever reason. I guess you could maybe get around this by only loading the script once during construction, like so:

```groovy
Map loadedScripts = [:]

@Before
@Override
void setUp() throws Exception {
    super.setUp()
    loadVarsScriptsToBinding()
}

private void loadVarsScriptsToBinding() {
    // Add bindings to all the "vars" scripts
    File varsDir = new File("vars")
    for (File f : varsDir.listFiles()) {
        if (!f.name.endsWith(".groovy")) {
            continue
        }
        if (!loadedScripts.contains(f.name)) {
            loadedScripts[f.name] = loadScript(f.path)
        }
        String fileName = f.name.replace(".groovy", "")
        binding.setVariable(fileName, loadedScripts[f.name].clone())
    }
}
```

Note that I haven't tried the above code, so it may or may not actually work. In the case that it does work, this is just duct taping over the actual problem, so it might be worth some time profiling your scripts to figure out why they take so long to be loaded. I don't think that the performance of `loadScript` is the culprit, for example on the largest pipeline library which I maintain, I counted 17 instances of `loadScript` in the unit tests, but the whole test suite runs in less than 3 seconds on my machine.

**Comment by mahmoud-ashi** [comment]
Thanks again @nre-ableton  for taking time to reply.
Regarding the `build.gradle`, I will do the further simplification although I remember having some issues previously when removing the Jenkins jars. I will try that again to see what is the issue.

Regarding the loading, I already tried what you suggested and mentioned it in my previous comment "_I also tried creating a static map which holds the script name and its loaded script and later on just add those again to the binding but this does not seem to work_". The only difference is that I did not use the `clone()` on it. I just tried it now with clone and seems to break all tests with an exception `java.lang.CloneNotSupportedException` since we are trying to clone a `Script` object.

Nevertheless, refactoring my shared lib now would be difficult as I am not the only user of it and many contribute to it as well. Additionally, my `vars` scripts do not contain a simple `call` method but instead they act as libraries with each script containing many helper methods like I stated in my first comment.

Regarding the profiling, I just surrounded each `loadScript` call with a timeout to see how long each takes, and some take ~2seconds to load 1300+ lines of code.

One thing I also tried is to calculate the dependency of each script on the others and only load the dependencies and ignore the rest but this seems to be a problem because the dependencies of the dependencies still need to be loaded as well.

I will try to experiment with some other test runners to see if that could resolve the issue because I feel that initing the `helper` could be one cause of the issue since loaded scripts are "registered" on the helper and its binding.

**Comment by nre-ableton** [comment]
> Regarding the loading, I already tried what you suggested and mentioned it in my previous comment "_I also tried creating a static map which holds the script name and its loaded script and later on just add those again to the binding but this does not seem to work_". The only difference is that I did not use the `clone()` on it. I just tried it now with clone and seems to break all tests with an exception `java.lang.CloneNotSupportedException` since we are trying to clone a `Script` object.

First, sorry for overlooking that statement in your previous comment. Anyways, I guess that object can't be cloned, which makes sense I suppose.
 
> Nevertheless, refactoring my shared lib now would be difficult as I am not the only user of it and many contribute to it as well. Additionally, my `vars` scripts do not contain a simple `call` method but instead they act as libraries with each script containing many helper methods like I stated in my first comment.

Well, at least having good unit tests in place will make refactoring much safer. :smile: 
 
> Regarding the profiling, I just surrounded each `loadScript` call with a timeout to see how long each takes, and some take ~2seconds to load 1300+ lines of code.
> 
> One thing I also tried is to calculate the dependency of each script on the others and only load the dependencies and ignore the rest but this seems to be a problem because the dependencies of the dependencies still need to be loaded as well.
> 
> I will try to experiment with some other test runners to see if that could resolve the issue because I feel that initing the `helper` could be one cause of the issue since loaded scripts are "registered" on the helper and its binding.

Good luck! BTW, do you think that this issue can be closed? Are you satisfied with the solution here so far?

**Comment by mahmoud-ashi** [comment]
Thanks for all your support @nre-ableton. I think I can close this ticket now as the remaining issue is not related. I will continue playing around with it and if I need help, I will open a new issue. Thanks again.
